cinder-lvm

lmlg opened the pull request:
Implements a test suite for the cinder-lvm subordinate charm, plus a few fixes here and there.

lmlg committed:
Reorganize code and add unit tests

The initial code for the charm had a few things missing. In addition,
unit tests weren't running, so a set was added, mocking the appropriate
charm functionality.

lmlg committed:
Reorganize code and add unit tests

The initial code for the charm had a few things missing. In addition,
unit tests weren't running, so a set was added, mocking the appropriate
charm functionality.

lmlg committed:
Update requirements

lmlg committed:
Update requirements

lmlg committed:
Test loopback devices

lmlg committed:
Ignore .swp files

lmlg committed:
Call base tearDown in tests

lmlg committed:
First commit

lmlg committed:
Link the README file

wolsen commented on src/lib/charm/openstack/cinder_lvm.py:
```diff
@@ -325,12 +337,12 @@ def _parse_block_device(block_device):
     name = 'cinder_lvm'
     release = 'ocata'
-    packages = []
+    packages = ['cinder-volume']
```
This is a [list of packages](https://opendev.org/openstack/charms.openstack/src/branch/master/charms_openstack/charm/classes.py#L1035) that should be installed, but the cinder-volume service shouldn't need to be installed here as it is a subordinate charm and must be paired with the cinder charm which will bring along cinder-volume with it.

wolsen commented on src/test-requirements.txt:
```diff
@@ -1,3 +1,6 @@
 # zaza
 git+https://github.com/openstack-charmers/zaza.git#egg=zaza
 git+https://github.com/openstack-charmers/zaza-openstack-tests.git#egg=zaza.openstack
+charmhelpers
```
The charmhelpers is typically defined in the wheelhouse.txt and should point to `git+https://github.com/juju/charm-helpers.git#egg=charmhelpers`.

wolsen commented on tox.ini:
```diff
@@ -1,61 +1,132 @@
-# Source charm: ./tox.ini
+# Classic charm (with zaza): ./tox.ini
```
This is a Reactive charm

wolsen commented on unit_tests/test_lib_charm_openstack_cinder_lvm.py:
```diff
@@ -12,29 +12,175 @@
-        self.patch_object(charmhelpers.core.hookenv, 'config')
+    @classmethod
+    def setUpClass(cls):
+        cls.DFL_CONFIG = {'block-device': '/dev/sdb', 'overwrite': False,
```
What does DFL_CONFIG mean?

lmlg replied:
Default configuration. Essentially, what should be configured for the charm before modifications.

wolsen commented on unit_tests/test_lib_charm_openstack_cinder_lvm.py:
```diff
@@ -12,29 +12,175 @@
+        self._config = self.DFL_CONFIG.copy()
+        lvm = self.LVM
+
+        def cf(key=NIL):
```
Any particular reason you're using a NIL object vs None here? None is fairly standard and its non-obvious to the casual reader why you have a NIL object rather than just using None.

lmlg replied:
Just the way I've always done it. I'll switch to `None` since it's less idiosyncratic.

wolsen commented on src/lib/charm/openstack/cinder_lvm.py:
```diff
@@ -290,7 +289,8 @@ def clean_storage(block_device):
         deactivate_lvm_volume_group(block_device)
-        remove_lvm_physical_volume(block_device)
+        subprocess.Popen(['pvremove', '-ff', block_device],
+            stdin=subprocess.PIPE).communicate(b'y\n')
```
This change should be made against the [charmhelpers](https://github.com/juju/charm-helpers) library instead and this should still use the remove_lvm_physical_volume method. We should fix the helpers library so all charms using this can benefit.

lmlg replied:
Agreed. This is a leftover of the change made so that the tests would run.

wolsen replied:
I don't see a corresponding PR against charmhelpers for this. Did you make one?

lmlg replied:
Done: https://github.com/juju/charm-helpers/pull/636

wolsen commented on src/tests/tests_cinder_lvm.py:
```diff
@@ -24,24 +24,52 @@
 
-class CinderlvmTest(test_utils.OpenStackBaseTest):
-    """Encapsulate lvm tests."""
+def with_conf(key, value):
```
This whole test class should be proposed against the [zaza-openstack-tests](https://github.com/openstack-charmers/zaza-openstack-tests) repository and not part of the charm itself.

wolsen replied:
It looks like this has been proposed to https://github.com/openstack-charmers/zaza-openstack-tests/pull/626

wolsen commented on src/tests/tests_cinder_lvm.py:
```diff
@@ -52,19 +80,42 @@ def test_cinder_config(self):
+            timeout=10)
 
-    def test_create_volume(self):
+    def _tst_create_volume(self):
```
This method is quite close to test_create_volume and could be a bit confusing. A different name for the function would be a bit more clear that its a utility function to do the create volume for the various permutations.

wolsen commented on src/lib/charm/openstack/cinder_lvm.py:
```diff
@@ -122,14 +136,12 @@ def configure_lvm_storage(block_devices, volume_group, overwrite=False,
                 new_devices.append(device)
-        elif (is_lvm_physical_volume(device)
-              and list_lvm_volume_group(device) != volume_group):
+        elif list_lvm_volume_group(device) != volume_group:
```
nice logic simplification for this if elif else block

wolsen commented:
I think the merging of the cinder-lvm-dev branch into your local branch is causing your git history to be weird here. Commits exist multiple times which makes it quite confusing when doing the merge/review. I think we want to clean this up before any kind of merge.

lmlg commented:
> I think the merging of the cinder-lvm-dev branch into your local branch is causing your git history to be weird here. Commits exist multiple times which makes it quite confusing when doing the merge/review. I think we want to clean this up before any kind of merge.

Looking at it, it does appear that something weird is going on. I think I'll create a new branch with a single commit and the changes proposed thus far. What do you think ?

wolsen commented:
> Looking at it, it does appear that something weird is going on. I think I'll create a new branch with a single commit and the changes proposed thus far. What do you think ?

A new branch will require a whole new PR and I think there's value in keeping the history of the PR here. I believe it can be resolved locally and then force pushed to your branch, though I'm not entirely sure how that affects the conversation. Worst case can put up a new PR, but I think its resolvable on this branch.

lmlg commented:
> > Looking at it, it does appear that something weird is going on. I think I'll create a new branch with a single commit and the changes proposed thus far. What do you think ?
> 
> A new branch will require a whole new PR and I think there's value in keeping the history of the PR here. I believe it can be resolved locally and then force pushed to your branch, though I'm not entirely sure how that affects the conversation. Worst case can put up a new PR, but I think its resolvable on this branch.

Alright, I'll rewrite git history as tidy as possible then.

lmlg committed:
Improve tests

lmlg committed:
Reinstate import

lmlg committed:
Improve the 'with_conf' decorator

The previous implementation was hardcoded to work with the 'cinder-lvm'
charm, and only set one option at a time. This commit makes it more
generic in both regards.

lmlg committed:
Add new bundles

Adds the bundles for functional tests for focal-{ussury,victoria,wallaby},
groovy-victoria and hirsute-wallaby.

lmlg committed:
Rewrite the README file

The original file was somewhat outdate. This new one is done according to
the standards adopted by Canonical.

wolsen commented on src/tests/tests_cinder_lvm.py:
```diff
@@ -82,7 +85,7 @@ def test_cinder_config(self):
             timeout=10)
 
-    def _tst_create_volume(self):
+    def _create_volume(self):
```
this is changed here, but it should be changed in the PR for zaza-openstack-tests and this whole class dropped from this PR please.

lmlg committed:
Remove the test file itself.

It was moved to the zaza tests repository, so it has no reason to
be here.

pmatulis commented on src/README.md:
```diff
@@ -1,140 +1,115 @@
+
+  juju add-relation cinder-lvm cinder
+
+## Actions
```
We do not include descriptions for the actions. Just a list. The intro paragraph explains how to get descriptions.

pmatulis commented on src/README.md:
```diff
@@ -1,140 +1,115 @@
+
+This charm's primary use is as a backend for the cinder charm. To do so, add a relation betweeen both charms:
+
+  juju add-relation cinder-lvm cinder
```
Please include explicit endpoints:

`juju add-relation cinder-lvm:<endpoint> cinder:<endpoint>`

pmatulis commented on src/README.md:
```diff
@@ -1,140 +1,115 @@
+Because this is a subordinate charm a relation will need to be added to the 'cinder' charm
+application to have the charm deployed on a machine.
+
+## Cinder-lvm as a backend for Cinder
```
Probably don't need this header

lmlg replied:
Got it, I was basing it on the example template, section 'Swift as backend for Glance'

pmatulis commented on src/README.md:
```diff
@@ -1,140 +1,115 @@
+
+This section covers common and/or important configuration options. See file `config.yaml` for the full list of options, along with their descriptions and default values. See the [Juju documentation][juju-docs-config-apps] for details on configuring applications.
+
+#### `alias`
```
Please only include those options that we really want the user to know about.

pmatulis commented on src/README.md:
```diff
@@ -1,140 +1,115 @@
-high-availability.
+# Overview
+
+Cinder-lvm is a charm that works as a subordinate for the cinder charm, implementing the LVM backend.
```
I try to avoid capitalising charm names. Maybe "The cinder-lvm charm..."

pmatulis replied:
If there are minimum requirements ( e.g. OpenStack release) for the charm to work please state them in the overview. See the [ovn-central](https://opendev.org/x/charm-ovn-central/src/branch/master/src/README.md) charm as a model.

pmatulis replied:
I like providing a very basic intro of the main ingredients (for the benefit of new users). Here it would be Cinder and LVM. See the [cinder-ceph](https://opendev.org/openstack/charm-cinder-ceph/src/branch/master/README.md) charm as a model.

pmatulis commented on src/README.md:
```diff
@@ -1,140 +1,115 @@
+
+Please report bugs on [Launchpad][lp-bugs-charm-cinder-lvm].
+
+<!-- LINKS -->
```
Ensure all necessary links are present and any unnecessary links are removed.

pmatulis reviewed (changes requested):
Thanks a lot for using the README template. Please see inline commentary.

lmlg committed:
Standardize charm

Add or move the files that are needed to make the charm compliant with
the standard defined by 'charm-proof'. This includes the config.yaml,
copyright and bundle files.

lmlg committed:
Improve the README file

Based on comments received, unused links were dropped, certain parts
were clarified and the file itself was streamlined.

lmlg committed:
Update release requirements and test parameter name

Change 'DFL_CONFIG' to 'DEFAULT_CONFIG'; furthermore, specify the
minimum Openstack requirement, in documentation and code.

pmatulis reviewed (commented):
README looks good!

freyes commented on metadata.yaml:
```diff
@@ -1,4 +1,4 @@
-name: cinder-lvm
+name: charm-cinder-lvm
```
this shouldn't carry the `charm-` prefix, if this is to make `charm proof` happy it's better to rename the working directory (e.g. `git clone https://github.com/openstack-charmers/charm-cinder-lvm.git cinder-lvm`).

freyes commented on src/README.md:
```diff
@@ -1,140 +1,99 @@
-high-availability.
+# Overview
+
+The cinder charm is the Openstack block storage (i.e: Volume) service, whereas the cinder-lvm charm works as a subordinate of cinder, implementing a backend based on the Linux Volume Manager (LVM).
```
I'm not sure if this is a written rule, but lines' width are 80 chars wide.

lmlg replied:
The readme template I was using had lines above 80 columns wide, so I thought it was OK.

freyes commented on src/tests/bundles/bionic-queens.yaml:
```diff
@@ -0,0 +1,50 @@
+    charm: cs:~openstack-charmers-next/keystone
+    num_units: 1
+    to:
+    - '1'
```
why not using different machines for each juju application?

lmlg replied:
No particular reason. The repo had already a bundle I was basing these off, so I went with what was already there.

freyes commented on src/tests/bundles/focal-ussuri.yaml:
```diff
@@ -0,0 +1,60 @@
+  mysql:
+    charm: cs:~openstack-charmers-next/percona-cluster
+    num_units: 1
+    series: bionic
```
is there any reason to not use focal?, for clouds running >=focal we use mysql-innodb-cluster , see https://github.com/openstack/charm-keystone/blob/master/tests/bundles/focal-ussuri.yaml#L27-L35

lmlg replied:
Mostly just for simplicity, and to have the bundles as similar as possible, but I can change them if needed.

wolsen replied:
Yeah, please change to mysql-innodb-cluster and focal as we use that everywhere else for focal.

freyes commented on src/tests/tests.yaml:
```diff
@@ -1,9 +1,9 @@
 configure:
   - zaza.openstack.charm_tests.keystone.setup.add_demo_user
 gate_bundles:
-  - xenial-ocata
```
cinder-lvm will support >=bionic-queens, we should drop the bundles for Xenial-

freyes commented on test-requirements.txt:
```diff
@@ -5,3 +5,6 @@ charms.reactive
 mock>=1.2
 coverage>=3.6
 git+https://github.com/openstack/charms.openstack.git#egg=charms-openstack
+charmhelpers
```
I believe we should be using `git+https://github.com/juju/charm-helpers.git#egg=charmhelpers` would be better for once this charm gets released we can switch to the `stable/*` branch

freyes commented on tox.ini:
```diff
@@ -1,61 +1,137 @@
+[testenv:build]
+basepython = python3
+commands =
+    charm build --log-level DEBUG -o {toxinidir}/build src {posargs}
```
I get this warning when running this target:

```
$ tox -e build
build create: /home/freyes/Projects/charms/openstack/cinder-lvm/.tox/build
build installdeps: -r/home/freyes/Projects/charms/openstack/cinder-lvm/test-requirements.txt
build installed: attrs==21.2.0,autopage==0.4.0,charmhelpers==0.20.22,charms.openstack @ git+https://github.com/openstack/charms.openstack.git@6c8eb01162edef5e3f8e3af6cb90905e8b2a3065,charms.reactive==1.4.1,cliff==3.9.0,cmd2==2.1.2,colorama==0.4.4,coverage==5.5,extras==1.0.0,fixtures==3.0.0,flake8==3.9.2,future==0.18.2,Jinja2==3.0.1,MarkupSafe==2.0.1,mccabe==0.6.1,mock==4.0.3,netaddr==0.8.0,netifaces==0.11.0,os-testr==2.0.0,pbr==5.6.0,prettytable==2.2.0,psutil==5.8.0,pyaml==21.8.3,pycodestyle==2.7.0,pyflakes==2.3.1,pyparsing==2.4.7,pyperclip==1.8.2,python-subunit==1.4.0,PyYAML==5.4.1,six==1.16.0,stestr==3.2.0,stevedore==3.4.0,Tempita==0.5.2,testtools==2.5.0,voluptuous==0.12.1,wcwidth==0.2.5
build run-test-pre: PYTHONHASHSEED='0'
build run-test: commands[0] | charm build --log-level DEBUG -o /home/freyes/Projects/charms/openstack/cinder-lvm/build src
WARNING: test command found but not installed in testenv
  cmd: /snap/bin/charm
  env: /home/freyes/Projects/charms/openstack/cinder-lvm/.tox/build
Maybe you forgot to specify a dependency? See also the allowlist_externals envconfig setting.

DEPRECATION WARNING: this will be an error in tox 4 and above!
```

And the build charm gets the wrong directory name:

```
$ ls build/
src
$ ls build/src/
actions       bin          copyright                exec.d  layer.yaml  LICENSE   metadata.yaml  README.md         templates              tests    version     wheelhouse.txt
actions.yaml  config.yaml  copyright.layer-options  hooks   lib         Makefile  reactive       requirements.txt  test-requirements.txt  tox.ini  wheelhouse
```

lmlg replied:
Made a blunder when updating some config parameters. I've fixed it now, should be all green.

freyes reviewed (changes requested):
unit tests fail when running `tox`.

```
======
Totals
======
Ran: 6 tests in 0.1384 sec.
 - Passed: 2
 - Skipped: 0
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 4
Sum of execute time for each test: 0.1909 sec.

==============
Worker Balance
==============
 - Worker 0 (2 tests) => 0:00:00.047537
 - Worker 1 (2 tests) => 0:00:00.099925
 - Worker 2 (1 tests) => 0:00:00.046217
 - Worker 3 (1 tests) => 0:00:00.009811
Test id                                                                                            Runtime (s)
-------------------------------------------------------------------------------------------------  -----------
unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_lvm_loopback_dev     0.086
unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_lvm_block_dev_none   0.046
unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_lvm_ephemeral_mount  0.020
unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_lvm_single_vg        0.017
unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_configuration        0.012
unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_base                 0.010
ERROR: InvocationError for command /home/freyes/Projects/charms/openstack/cinder-lvm/.tox/py3/bin/stestr run --slowest (exited with code 1)
```

I still need to take more time to review the unit tests, but these are the things I found on the first round.

Commit:
Fix failing unit tests

After the last commit, when some configuration parameters were
changed, some tests broke.

Commit:
Drop the 'charm-' prefix in the name category for the metadata file

Commit:
Update bundles, dropping unsupported versions

Commit:
Point charmhelpers to the correct repo in test-requirements.txt

lmlg commented:
> unit tests fail when running `tox`.
> 
> ```
> ======
> Totals
> ======
> Ran: 6 tests in 0.1384 sec.
>  - Passed: 2
>  - Skipped: 0
>  - Expected Fail: 0
>  - Unexpected Success: 0
>  - Failed: 4
> Sum of execute time for each test: 0.1909 sec.
> 
> ==============
> Worker Balance
> ==============
>  - Worker 0 (2 tests) => 0:00:00.047537
>  - Worker 1 (2 tests) => 0:00:00.099925
>  - Worker 2 (1 tests) => 0:00:00.046217
>  - Worker 3 (1 tests) => 0:00:00.009811
> Test id                                                                                            Runtime (s)
> -------------------------------------------------------------------------------------------------  -----------
> unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_lvm_loopback_dev     0.086
> unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_lvm_block_dev_none   0.046
> unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_lvm_ephemeral_mount  0.020
> unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_lvm_single_vg        0.017
> unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_configuration        0.012
> unit_tests.test_lib_charm_openstack_cinder_lvm.TestCinderLVMCharm.test_cinder_base                 0.010
> ERROR: InvocationError for command /home/freyes/Projects/charms/openstack/cinder-lvm/.tox/py3/bin/stestr run --slowest (exited with code 1)
> ```
> 
> I still need to take more time to review the unit tests, but these are the things I found on the first round.

Made a blunder when updating some config parameters. I've fixed it now, should be all green.

Commit:
Cleanup tox.ini file

This change makes the built charm be 'cinder-lvm' instead of source,
and sets the standard environment variables for charm building.

Commit:
Reinstate files under src directory

A previous commit erroneously deleted them, but they are needed for
functional tests to properly run.

Commit:
Remove unnecesary environment vars from tox.ini file

Since they can be set individually, avoid polluting the environment
with unneeded variables

lmlg committed:
Modernize the focal/groovy bundles

Instead of hardcoding the series so that mysql/percona continues
working, move to mysql-innodb-cluster. Furthermore, use a separate
machine for every application.

lmlg committed:
Update the tests.yaml file

lmlg committed:
Limit smoke testing to a single bundle

lmlg committed:
Standardize bundle files

Follow the blueprint defined by the cinder charm and use the same
format for the functional test bundles

lourot commented:
@lmlg can be closed because repo has moved?

lmlg commented:
> @lmlg can be closed because repo has moved?

Yep, forgot about this. Closing it now.
